File: Abstraction/python/Interface.py
from abc import ABC, abstractmethod

# Define the Animal interface
class Animal(ABC):
    # Define the abstract method
    @abstractmethod
    def speak(self):
        pass

# Dog class implementing the Animal interface
class Dog(Animal):
    def speak(self):
        return "Woof!"

# Cat class implementing the Animal interface
class Cat(Animal):
    def speak(self):
        return "Meow!"

# Usage example
def animal_sound(animal: Animal):
    print(animal.speak())

# Instantiate Dog and Cat
dog = Dog()
cat = Cat()

# Call the animal_sound function with Dog and Cat objects
animal_sound(dog)  # Outputs "Woof!"
animal_sound(cat)  # Outputs "Meow!"

File: Abstraction/CPP/Interface.cpp
#include <iostream>

// Define the Animal interface
class Animal {
public:
    // Pure virtual function for speaking
    virtual void speak() const = 0;

    // Virtual destructor for proper cleanup
    virtual ~Animal() = default;
};

// Dog class implementing the Animal interface
class Dog : public Animal {
public:
    void speak() const override {
        std::cout << "Woof!" << std::endl;
    }
};

// Cat class implementing the Animal interface
class Cat : public Animal {
public:
    void speak() const override {
        std::cout << "Meow!" << std::endl;
    }
};

int main() {
    // Create Dog and Cat objects
    Dog myDog;
    Cat myCat;

    // Call speak method on each animal
    myDog.speak(); // Outputs "Woof!"
    myCat.speak(); // Outputs "Meow!"

    return 0;
}
